### PR TITLE
Major Package Updates and SQLAlchemy 2

### LINKED ISSUE #108: Safety run failed
https://github.com/PyBites-Open-Source/karmabot/actions/runs/4073711282

@pogross maybe time to upgrade some of the packages ...

### PR DESCRIPTION
**Pull request by pogross**
Fixes #108 and  #107 

- Switch to Python 3.10 and 3.11 as supported versions
- Updates all packages to latest versions
- Updates Github Actions
- Migration from SQLAlchemy 1.4.20 to 2.0.1
- Update of `pre-commit-config.yaml` to latest hooks
- New poetry `pyproject.toml` format with `[tool.poetry.dependencies]` and `[tool.poetry.group.dev.dependencies]`

Please do some manual testing ;) For me all database related commands worked :)

### CODE REVIEW

**Comment by pogross** [comment]
Somehow the old Actions for 3.8 and 3.9 are dangling. We might need to force merge when review is done 🔨

**Comment by pogross** [comment]
@pybob ping ;)

**Comment by pogross** [comment]
@bbelderbos  ping

File: pyproject.toml [hunk 1]
```diff
@@ -92,27 +92,27 @@ exclude = '''
...
+ignore_missing_imports = true
 mypy_path = "src,tests"
 pretty = true
 show_column_numbers = true
 show_error_codes = true
 show_error_context = true
-ignore_missing_imports = true
 
 [[tool.mypy.overrides]]
+ignore_missing_imports = true
```
Comment (bbelderbos) [inline pyproject.toml:110]: seems defined twice?

File: pyproject.toml [hunk 2]
```diff
@@ -92,27 +92,27 @@ exclude = '''
...
 module = [
   "freezegun",
-  "requests"
+  "requests",
```
Comment (bbelderbos) [inline pyproject.toml:113]: pinned above, do we also need them here?

File: src/karmabot/bot.py [hunk 3]
```diff
@@ -71,22 +71,22 @@ def compile_command_pattern(commands: Dict[str, Callable]) -> re.Pattern:
     command_words = commands.keys()
     all_commands = "|".join(command_words)
 
-    full_commands = fr"^<@{KARMABOT_ID}>\s*({all_commands})(\s.*)?$"
+    full_commands = rf"^<@{KARMABOT_ID}>\s*({all_commands})(\s.*)?$"
```
Comment (bbelderbos) [inline src/karmabot/bot.py:74]: yep liking this better

File: src/karmabot/bot.py [hunk 4]
```diff
@@ -185,8 +185,8 @@ def reply_commands(message, say):  # noqa
         channel_id = message["channel"]
         text = message["text"]
         channel_type = message["channel_type"]
-    except KeyError as exc:
-        logging.error("reply_commands error! Message was: %s", message, exc_info=exc)
+    except KeyError:
+        logging.exception("reply_commands error! Message was: %s", message)
```
Comment (bbelderbos) [inline src/karmabot/bot.py:189]: nice

File: src/karmabot/commands/note.py [hunk 5]
```diff
@@ -59,7 +61,9 @@ def _del_note(text: str, user: KarmaUser) -> str:
         return f"Sorry {user.username}, it seems you did not provide a valid id."
 
     with database.session_manager() as session:
-        query = session.query(KarmaNote).filter_by(id=note_id, user_id=user.user_id)
+        query = session.execute(
```
Comment (bbelderbos) [inline src/karmabot/commands/note.py:64]: looks more like sqlmodel :)

File: src/karmabot/karma.py [hunk 6]
```diff
@@ -100,9 +101,13 @@ def _save_transaction(self, points):
         self.session.commit()
 
         finished_transaction = (
-            self.session.query(KarmaTransaction)
-            .order_by(KarmaTransaction.id.desc())
-            .first()
+            self.session.execute(
+                select(KarmaTransaction).order_by(KarmaTransaction.id.desc())
+            ).first()
+            # SQLAlchemy migration
```
Comment (bbelderbos) [inline src/karmabot/karma.py:107]: remove?

File: src/karmabot/settings.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 from dotenv import load_dotenv
 
 
-def get_env_var(env_var: str, default: str = None) -> str:
+def get_env_var(env_var: str, default: str | None = None) -> str:
```
Comment (bbelderbos) [inline src/karmabot/settings.py:9]: advantage for >= 3.10 (I believe), nice

**Pull request merged**